src/libaktualizr/primary/aktualizr.cc

```diff
@@ -28,25 +23,30 @@ Aktualizr::Aktualizr(Config &config) : config_(config) {
   urandom.close();
   std::srand(seed);  // seeds pseudo random generator with random number
   LOG_TRACE << "... seeding complete in " << timer;
+
+  commands_channel_ = std::make_shared<command::Channel>();
+  events_channel_ = std::make_shared<event::Channel>();
+  sig_ = std::make_shared<boost::signals2::signal<void(std::shared_ptr<event::BaseEvent>)>>();
 }
 
 int Aktualizr::run() {
-  std::shared_ptr<command::Channel> commands_channel{new command::Channel};
-  std::shared_ptr<event::Channel> events_channel{new event::Channel};
-
-  EventsInterpreter events_interpreter(config_, events_channel, commands_channel);
+  EventsInterpreter events_interpreter(config_, events_channel_, commands_channel_, sig_);
 
   // run events interpreter in background
   events_interpreter.interpret();
 
   std::shared_ptr<INvStorage> storage = INvStorage::newStorage(config_.storage);
   storage->importData(config_.import);
-  HttpClient http;
-  ReportQueue report_queue(config_, http);
-  Uptane::Manifest uptane_manifest{config_, storage};
-  Bootloader bootloader(config_.bootloader);
-  SotaUptaneClient uptane_client(config_, events_channel, uptane_manifest, storage, http, bootloader, report_queue);
-  uptane_client.runForever(commands_channel);
+
+  std::shared_ptr<SotaUptaneClient> uptane_client =
+      SotaUptaneClient::newDefaultClient(config_, storage, events_channel_);
+  uptane_client->runForever(commands_channel_);
 
   return EXIT_SUCCESS;
 }
+
+void Aktualizr::sendCommand(std::shared_ptr<command::BaseCommand> command) { *commands_channel_ << command; }
+
+void Aktualizr::setSignalHandler(std::function<void(std::shared_ptr<event::BaseEvent>)> &handler) {
```

Comment:
It could be useful for the api user to remove a signal handler.

One possibility: return the `boost::signal2::connection` returned by `connect()`, so that `disconnect()` can be called from there.

Reply:
How bad would it be to directly expose the boost::signals2::signal object?  We'd be binding our API to boost, but on the other hand we'd get all the logic like automatic disconnection for free.

Reply:
It's an option. We are already bound to boost if I return the connection object (as I have now changed it to do). Sharing the signal object would expose `std::shared_ptr<boost::signals2::signal<void(std::shared_ptr<event::BaseEvent>)>>` instead of what I currently have (`boost::signals2::connection setSignalHandler(std::function<void(std::shared_ptr<event::BaseEvent>)>& handler);`), which is already a bit of a handful. I was considering ways to `typedef`/`using` that into something nicer, but I'm not sure what that would be that would actually make anyone's life easier. (Doing it for the function object is an option, but honestly it is probably better to know upfront what the required input and outputs of the handler are.)